[MIG] resource_booking: migrate to v13

yajo opened the pull request:
- Standard v13 changes.
- Some onchanges removed in favor of computes.

@Tecnativa TT30987

Commit:
[ADD] resource_booking: new app to manage bookings
This module adds a new app to allow you to book resource combinations in given
schedules.

Example use cases:

* Management of consultations in a clinic.
* Salesman appointments.
* Classroom and projector reservations.
* Hotel room booking.

Among the things you can do:

* Specify the type of booking, which includes a calendar of availability.
* Specify which resources can be booked together. All of them must be free to be booked.
* Place pending bookings, effectively giving permissions to someone to see the availability calendar and choose one slot.
* Partners can do that from their portals.
* If a partner has no user, he can still do the same via a tokenized URL.
* Backend users can also do that from the backend.
* Booking lifecycle with computed states.
* Automatic meeting creation and deletion.
* Automatic conflict detection.
* Deadline to block modifications.

@Tecnativa TT28201

oca-travis committed:
[UPD] Update resource_booking.pot

OCA-git-bot committed:
[UPD] README.rst

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (190 of 190 strings)

Translation: calendar-12.0/calendar-12.0-resource_booking
Translate-URL: https://translation.odoo-community.org/projects/calendar-12-0/calendar-12-0-resource_booking/es/

Commit:
[IMP] resource_booking: filter/group by date

TT29508

oca-travis committed:
[UPD] Update resource_booking.pot

OCA-git-bot committed:
resource_booking 12.0.1.1.0

Commit:
[FIX] resource_calendar: ignore past or unconfirmed bookings when updating calendars

Without this patch, users couldn't change a calendar schedule if there were past or unconfirmed bookings that wouldn't fit in it.

Excluding those bookings from the check fixes the situation.

We also check that, to confirm a booking, it must fit in the calendar (because now it can happen that, in the time that has passed since the booking was scheduled until it is confirmed, the calendar changes).

@Tecnativa TT29509

oca-travis committed:
[UPD] Update resource_booking.pot

OCA-git-bot committed:
resource_booking 12.0.1.1.1

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: calendar-12.0/calendar-12.0-resource_booking
Translate-URL: https://translation.odoo-community.org/projects/calendar-12-0/calendar-12-0-resource_booking/

Commit:
[FIX] resource_booking: always notify in resource TZ

The notifications emitted to the resource booking requester must always be in the same TZ as the resource booking itself.

For example, if you book one hotel room in the other side of the world, a notification in your own TZ is confusing.

Besides, res.partner created from website_sale are created with `tz=False`, making it even more confusing.

@Tecnativa TT30331

OCA-git-bot committed:
resource_booking 12.0.1.1.2

joao-p-marques committed:
[FIX] resource_booking: Limit constraint only to future bookings

The constraint that checks the schedule of a resource booking is currently being applied to all the bookings, including past ones.
As the resource combination or associated calendars might change regularly and trigger a recomputation of this, such change might take a very long time.
Plus, the calendar restrictions might change, trigger a recompute of the constraint and detect bookings that can't be assigned, which makes no sense when they already happened.

This applies it only to future bookings, ignoring past ones.

TT30478

OCA-git-bot committed:
resource_booking 12.0.1.1.3

Commit:
[IMP] resource_booking: black, isort, prettier

Commit:
[MIG] resource_booking: migrate to v13

- Standard v13 changes.
- Some onchanges removed in favor of computes.

@Tecnativa TT30987

sergio-teruel commented on resource_booking/models/resource_booking.py:
```diff
@@ -0,0 +1,531 @@
+        ),
+    ]
+
+    active = fields.Boolean(index=True, default=True)
```
Remove index from boolean field

yajo replied:
Why? This is used almost always to filter values... 🤔

pedrobaeza replied:
https://stackoverflow.com/a/42972924

Short answer: it doesn't benefit

sergio-teruel commented:
Please review travis..

Commit:
[IMP] resource_booking: simplified time format

It's very unlikely that you need to book resources in a seconds-based precision.

This simplifies portal UI. Still, declared as a variable in case you need some customizations downstream.

@Tecnativa TT31063

chienandalu commented on resource_booking/README.rst:
```diff
@@ -0,0 +1,205 @@
+
+When someone is a manager, he will have access to *Resource Bookings >
+Configuration*, where he will be able to configure resources, leaves and
+schedules. This menu is just provided as a commodity. However, if you want to
```
Maybe you meant *amenity*?
```suggestion
schedules. This menu is just provided as a ?. However, if you want to
```

yajo replied:
I think "shortcut" fits better.

chienandalu reviewed (commented):
Trying to make a functional test a came into this error:

<details name="error">
  <summary>Traceback</summary>
<pre>Odoo Server Error

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 624, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 669, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/.repo_requirements/odoo/addons/web/controllers/main.py", line 1339, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/.repo_requirements/odoo/addons/web/controllers/main.py", line 1331, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 387, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 374, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/models.py", line 6251, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/.repo_requirements/odoo/odoo/models.py", line 6030, in _onchange_eval
    method_res = method(self)
  File "/home/odoo/build/OCA/calendar/resource_booking/models/resource_booking.py", line 340, in _onchange_start_fill_stop
    self.stop = self.start and self.start + timedelta(hours=self.type_id.duration)
  File "/.repo_requirements/odoo/odoo/fields.py", line 1115, in __set__
    new_records.modified([self.name])
  File "/.repo_requirements/odoo/odoo/models.py", line 5837, in modified
    for field, records, create in tocompute:
  File "/.repo_requirements/odoo/odoo/models.py", line 5891, in _modified_triggers
    records = self[invf.name]
  File "/.repo_requirements/odoo/odoo/models.py", line 5733, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/.repo_requirements/odoo/odoo/fields.py", line 2365, in __get__
    return super().__get__(records, owner)
  File "/.repo_requirements/odoo/odoo/fields.py", line 1031, in __get__
    self.compute_value(recs)
  File "/.repo_requirements/odoo/odoo/fields.py", line 1145, in compute_value
    records._compute_field_value(self)
  File "/.repo_requirements/odoo/odoo/models.py", line 4005, in _compute_field_value
    getattr(self, field.compute)()
  File "/home/odoo/build/OCA/calendar/resource_booking/models/resource_booking.py", line 154, in _compute_combination_id
    one.combination_id = one._get_best_combination()
  File "/home/odoo/build/OCA/calendar/resource_booking/models/resource_booking.py", line 397, in _get_best_combination
    availability = self._get_intervals(*desired_interval, combination)
  File "/home/odoo/build/OCA/calendar/resource_booking/models/resource_booking.py", line 454, in _get_intervals
    result &= combinations._get_intervals(start_dt, end_dt)
  File "/home/odoo/build/OCA/calendar/resource_booking/models/resource_booking_combination.py", line 81, in _get_intervals
    combination_intervals &= calendar._work_intervals(start_dt, end_dt, res)
  File "/.repo_requirements/odoo/addons/resource/models/resource.py", line 474, in _work_intervals
    start_dt, end_dt, resources=resource, domain=domain, tz=tz
  File "/.repo_requirements/odoo/addons/resource/models/resource.py", line 486, in _work_intervals_batch
    leave_intervals = self._leave_intervals_batch(start_dt, end_dt, resources, domain, tz=tz)
  File "/home/odoo/build/OCA/calendar/resource_booking/models/resource_calendar.py", line 109, in _leave_intervals_batch
    self.env.context["analyzing_booking"],
  File "/.repo_requirements/odoo/addons/resource/models/resource.py", line 102, in __or__
    return Intervals(chain(self._items, other._items))
  File "/.repo_requirements/odoo/addons/resource/models/resource.py", line 85, in __init__
    append((start, value, recses[0].union(*recses)))
  File "/.repo_requirements/odoo/odoo/models.py", line 5664, in union
    raise TypeError("Mixing apples and oranges: %s.union(%s)" % (self, arg))
TypeError: Mixing apples and oranges: calendar.event(8,).union(resource.calendar.leaves(1,))</pre>
</details>


Steps:

- I created a resource booking combination with the employee Mitchel Admin as resource
- I created a resource booking type with that combination
- Then just try to place a booking with that resource booking type :boom:

yajo commented:
I weren't able to reproduce https://github.com/OCA/calendar/pull/32#pullrequestreview-709264021, but based on the traceback I think it should be fixed now.

joao-p-marques commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+cssselect
+freezegun
```
This is just a doubt, not a real issue:

If this is a test requirement, should it go to `test-requirements.txt`? Also, is it correct to have it in `external_dependencies` in the manifest if it is only used for the tests?

yajo replied:
> If this is a test requirement, should it go to `test-requirements.txt`?

Indeed, fixed.

> Also, is it correct to have it in `external_dependencies` in the manifest if it is only used for the tests?

I'm not sure about this. I couldn't find any guidelines in https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#external-dependencies and Odoo doesn't have a way to tell if a dependency is general or just for testing.

As for us, in doodba, it doesn't make a difference, because all test or non-test dependencies are installed anyways, I found it more useful to add it to the manifest, so CI tells if the module failure comes from a missing dependency. But I might be wrong... 🤷🏼‍♂️ @pedrobaeza?

pedrobaeza replied:
Well, I don't want to impose my criteria in this, but for me, it shouldn't be on external_dependencies, as it's only needed for tests, not for general install.

joao-p-marques replied:
Thanks for the insights :+1: 

@Yajo I see you moved the entire file to `test-requirements.txt`. Is `cssselect` only for the tests as well? I thought only `freezegun` was...

yajo replied:
In this case it would be a bit cumbersome to move them out of the manifest, because `cssselect` isn't imported into the tests. Instead, the features it adds implicitly to `lxml` are used. See https://github.com/lxml/lxml/blob/02a49b1d6ad177c948652f8b4d72aa0e2b386b89/CHANGES.txt#L1638-L1642

Also, if we remove the dependencies from the manifest, we wouldn't be able to specify valid version constraints, if/when any. Or, if we did, that info would be decontextualized from the module.

🤔 ~~I think I'll just open~~ I opened one of those issues that Odoo will never fix in https://github.com/odoo/odoo/issues/74067, but just to keep track of this problem. But as long as we don't have a better place to put the dependencies, I think that putting them in the manifest will help more than harm.

Well, this is how it was before migrating anyway. Let's move on. 😄

pedrobaeza replied:
On v14, freezegun is included in Odoo core.

joao-p-marques reviewed (approved):
Code and functional review 👍 

Just a small doubt, otherwise LGTM

pedrobaeza commented:
/ocabot merge nobump
